README.md
# Nox-Browser
This is a basic web browser created using Python and PyQt5.
It hosts a special feature of reference mode that can be used to 
record the websites one has visited during a particular chosen period of time. 

Note: One has to install PyQt5, QWebEngine, datetime etc packages before running the code. Nox.py is the main file.
## Home Page
![HomePage](https://user-images.githubusercontent.com/48985763/83984924-e73b6d80-a954-11ea-941d-42b57b968426.png)
## Tab Page/Search Page
![TabPage](https://user-images.githubusercontent.com/48985763/83985009-32558080-a955-11ea-9f0c-031bcef517e8.png)
## Tab 2
![AnotherTab](https://user-images.githubusercontent.com/48985763/83985026-44cfba00-a955-11ea-90a6-6ed63540125f.PNG)
## Option Button
![OptionsPage](https://user-images.githubusercontent.com/48985763/83985046-54e79980-a955-11ea-8d9c-2d79b8c40c74.png)


History.txt
  [ 2020-06-06 06:05:13 ]  https://www.researchgate.net/

  [ 2020-06-06 06:05:26 ]  https://www.researchgate.net/login?_sg=NXEZDYSBUETXDsBV-CBFZo6izxWCvjJ_YLLiElCAK1njMRGjYp92iMApw9NWWDsG1xivYI-7vfalDITgHzWmhQ

  [ 2020-06-06 06:05:36 ]  https://accounts.google.com/signin/oauth?scope=profile+email&state=28154b877858b7afc1f20d36bf173466&redirect_uri=https://www.researchgate.net/connector/google/connected&response_type=code&client_id=757030871092-9tvmq9ga1e7q37dmfqc4nl2nrer96ecv.apps.googleusercontent.com&o2v=1&as=drKpKvIWBSoe4HILb-ZzWA

  [ 2020-06-06 06:05:39 ]  https://accounts.google.com/signin/oauth/identifier?scope=profile email&state=28154b877858b7afc1f20d36bf173466&redirect_uri=https%3A%2F%2Fwww.researchgate.net%2Fconnector%2Fgoogle%2Fconnected&response_type=code&client_id=757030871092-9tvmq9ga1e7q37dmfqc4nl2nrer96ecv.apps.googleusercontent.com&o2v=1&as=drKpKvIWBSoe4HILb-ZzWA&flowName=GeneralOAuthFlow

  [ 2020-06-06 06:06:12 ]  https://accounts.google.com/signin/v2/challenge/pwd?scope=profile email&state=28154b877858b7afc1f20d36bf173466&redirect_uri=https%3A%2F%2Fwww.researchgate.net%2Fconnector%2Fgoogle%2Fconnected&response_type=code&client_id=757030871092-9tvmq9ga1e7q37dmfqc4nl2nrer96ecv.apps.googleusercontent.com&o2v=1&as=drKpKvIWBSoe4HILb-ZzWA&flowName=GeneralOAuthFlow&cid=1&navigationDirection=forward&TL=AM3QAYaJceGhEp9nrwWlY2H-yqaVZw7v6WgrHQUFfyfrPE736fs4yuFtwWEv1RqS

  [ 2020-06-06 06:06:25 ]  https://accounts.google.co.in/accounts/SetSID?ssdc=1&sidt=ALWU2cuFUsEcLswel2Kj6bQljSIsK%2Fomn5nyb0dwP1kjYgmXEcORhHUtmY1cNmz0wY5qyovdSwl8rVHXKJxD53kXe2WDA1Xpa9%2BmlSSCMYuVGdvB%2Ba6VTEq9zJmhlKuvJHLYnYLntDwpKO20bdvd3KmpkqkyPQtwcVwSAz6Zh9Mr2sT359Aj8SZQNnNIJsumq5bYoKu5fr0%2FRKQfG6%2FTSM%2BUPGNFGw7RNXyzttZ%2BnFwtFEtzXhr6uNv6Zly7kWkMvmgHMwLsQhwgX2iiROk6phFfituWAFAdnasjBxhMz7fDIRRn8LPk9oc1aT4SV9FAy4L8V3ihvx2KhP6BhEpaNny749yIFfupJyIjsIld1DEPoOaotCNRqTA%3D&continue=https%3A%2F%2Faccounts.google.com%2Fsignin%2Foauth%2Fconsent%3Fauthuser%3D0%26part%3DAJi8hANYf5uM8Ka2SN3OOwd0oG3rtw_a6lQwg52FXwtAXb9uxOtl6Mnpg81RBWRnyORZC0h5CQmbz9EHAmXDwFK5vf5u1tAlNNrDfqOMi6kQaIj1j8-Dc6f9BpM9R7G863uaU8BeLb3sCcmWR6lIxyyiFXCgQXCFNbnfP1k5BlQNWFXUCDg309jQtHf8pjSOfnD7uXzyGrv1cAhI2eCZBaql99wzMf9WMIpqUIJxd2uDRTZbeEjrdwHKhtaNS4ovOgHG4HFqws3nOqK-c3qZIgZN5wC98h3rpe216Dh3XNvSd8PBHLXCxSpuYJX6CivtFwac8gu4khzAhhyvg3rLj7QaH9yPemvnj08C_k4juBET2e8KfKmi0MQCz5dONMLKCDeukT7TSePBiPziCGYugA4IIetBotHKvXeznNt7uZsTtnA-zVmW0ShqM04qODYV08bPQLlx_hC1%26as%3DdrKpKvIWBSoe4HILb-ZzWA%26rapt%3DAEjHL4MiZYtinrL-_9DQVE-ioqay5zHx6Mgxdp5uxtfQQoitvHvGIk_3NeN2Slz5hxVndnSd_1v2bn-v29U-0vrrKV9kaH1B-Q%23&tcc=1

  [ 2020-06-06 06:06:25 ]  https://accounts.google.co.in/accounts/SetSID

  [ 2020-06-06 06:06:28 ]  https://www.researchgate.net/signup.SignUp.html#

  [ 2020-06-06 06:06:44 ]  https://www.researchgate.net/connectors.ConnectorLoginFailed.html?dbw=true&networkName=Google&errorCode=0

  [ 2020-06-06 06:06:45 ]  https://www.researchgate.net/connectors.ConnectorLoginFailed.html?networkName=Google&errorCode=0

  [ 2020-06-06 06:06:50 ]  http://www.google.com

  [ 2020-06-06 06:06:50 ]  http://www.google.com/

  [ 2020-06-06 06:06:52 ]  https://www.google.com/

  [ 2020-06-06 06:07:28 ]  https://www.google.com/search?source=hp&ei=IeXaXofZB-rH4-EPmKui8AY&q=embedded+systems+pdf&oq=embedded+systems+pdf&gs_lcp=CgZwc3ktYWIQAzICCAAyAggAMgIIADICCAAyAggAMgYIABAWEB4yBggAEBYQHjIGCAAQFhAeMgYIABAWEB4yBggAEBYQHjoFCAAQgwE6BQgAELEDOgoIABCxAxBGEPkBUIHdAVinhQJg3YgCaABwAHgAgAHxBogB4C2SAQ0wLjQuOS4zLjEuMS4xmAEAoAEBqgEHZ3dzLXdpeg&sclient=psy-ab&ved=0ahUKEwjH7fKV-evpAhXq4zgGHZiVCG4Q4dUDCAc&uact=5

  [ 2020-06-06 06:08:33 ]  https://www.google.com/search?ei=RuXaXoigEvic4-EP9ICBqA4&q=coding+ninjas&oq=coding+ninjas&gs_lcp=CgZwc3ktYWIQAzIFCAAQsQMyAggAMgIIADICCAAyAggAMgIIADICCAAyAggAMgIIADICCAA6BAgAEEM6BwgAELEDEEM6BQgAEIMBOgUIABCRAlCi0QNY9-UDYPjqA2gAcAB4AIABmwSIAeUbkgELMC41LjQuMS4yLjGYAQCgAQGqAQdnd3Mtd2l6&sclient=psy-ab&ved=0ahUKEwiI28-n-evpAhV4zjgGHXRAAOUQ4dUDCAw&uact=5

  [ 2020-06-06 06:08:36 ]  https://www.codingninjas.com/

  [ 2020-06-06 06:09:46 ]  http://google.com/

  [ 2020-06-06 06:09:47 ]  https://www.google.com/

  [ 2020-06-06 06:09:54 ]  https://mail.google.com/mail/u/0/?ogbl#

  [ 2020-06-06 06:09:57 ]  https://mail.google.com/mail/u/0/?ogbl#inbox

  [ 2020-06-06 06:10:05 ]  https://mail.google.com/mail/u/0/?ogbl#inbox/FMfcgxwHNghCWMGGnKGCLkSTrcMPLLrL

  [ 2020-06-06 06:10:18 ]  https://mail.google.com/mail/u/0/?ogbl#inbox

  [ 2020-06-06 06:10:42 ]  http://www.google.com

  [ 2020-06-06 06:10:42 ]  http://www.google.com/

  [ 2020-06-06 06:10:44 ]  https://www.google.com/

  [ 2020-06-06 06:11:32 ]  https://www.google.com/

  [ 2020-06-06 06:11:41 ]  https://accounts.google.com/Logout?hl=en&continue=https://www.google.com/&timeStmp=1591404090&secTok=.AG5fkS_e8q5_HuzvqMaM5IgiWQlZtzFn9A

  [ 2020-06-06 06:11:41 ]  https://accounts.google.com/Logout

  [ 2020-06-06 06:11:41 ]  https://www.google.com/

  [ 2020-06-06 06:11:44 ]  https://accounts.google.com/ServiceLogin?hl=en&passive=true&continue=https://www.google.com/

  [ 2020-06-06 06:11:47 ]  https://accounts.google.com/ServiceLogin/signinchooser?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2F&flowName=GlifWebSignIn&flowEntry=ServiceLogin

  [ 2020-06-06 06:11:51 ]  https://accounts.google.com/ServiceLogin/signinchooser?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2F&flowName=GlifWebSignIn&flowEntry=AddSession

  [ 2020-06-06 06:11:51 ]  https://accounts.google.com/ServiceLogin/identifier?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2F&flowName=GlifWebSignIn&flowEntry=AddSession

  [ 2020-06-06 06:12:06 ]  https://accounts.google.com/ServiceLogin/deniedsigninrejected?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2F&flowName=GlifWebSignIn&flowEntry=AddSession

  [ 2020-06-06 06:12:09 ]  https://accounts.google.com/ServiceLogin/identifier?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2F&flowName=GlifWebSignIn&flowEntry=AddSession


option_window_class2.py
# This is the class used for creating the option dialog box.
# An object of the class is created when the option button is invoked




from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_Dialog(QtWidgets.QDialog):
    def __init__(self, tab_no):
        super().__init__()

        self.setObjectName("Dialog")


        self.setMinimumSize(QtCore.QSize(740, 478))
        self.setMaximumSize(QtCore.QSize(740, 478))

        self.Incog_btn = QtWidgets.QPushButton(self)
        self.Incog_btn.setGeometry(QtCore.QRect(0, 130, 241, 61))
        font = QtGui.QFont()
        font.setFamily("Cambria")
        font.setPointSize(14)
        self.Incog_btn.setFont(font)
        self.Incog_btn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.Incog_btn.setObjectName("Incog_btn")
        if tab_no > 1:
            self.Incog_btn.setEnabled(False)




        self.Bookmarks_btn = QtWidgets.QPushButton(self)
        self.Bookmarks_btn.setGeometry(QtCore.QRect(0, 10, 241, 61))
        font = QtGui.QFont()
        font.setFamily("Cambria")
        font.setPointSize(14)
        self.Bookmarks_btn.setFont(font)
        self.Bookmarks_btn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.Bookmarks_btn.setShortcut("")
        self.Bookmarks_btn.setObjectName("Bookmarks_btn")


        self.Ref_btn = QtWidgets.QPushButton(self)
        self.Ref_btn.setGeometry(QtCore.QRect(0, 190, 241, 61))
        font = QtGui.QFont()
        font.setFamily("Cambria")
        font.setPointSize(14)
        self.Ref_btn.setFont(font)
        self.Ref_btn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.Ref_btn.setObjectName("Ref_btn")
        if tab_no > 1:
            self.Ref_btn.setEnabled(False)



        self.show_area = QtWidgets.QTextEdit(self)
        self.show_area.setGeometry(QtCore.QRect(240, 10, 491, 461))
        font = QtGui.QFont()
        font.setFamily("Baskerville Old Face")
        font.setPointSize(12)
        self.show_area.setFont(font)
        self.show_area.setAutoFillBackground(False)
        self.show_area.setAutoFormatting(QtWidgets.QTextEdit.AutoBulletList)
        self.show_area.setReadOnly(True)
        self.show_area.setCursorWidth(0)
        self.show_area.setPlaceholderText("1.) Bookmarks - Contains links saved\n\n2.) History - Contains history of previous urls visited\n\n3.) Reference - A special feature to record history when this feature is on. On turning it off, it display the recorded history.\n\n4.) Incognito - Search without saving browser history\n\n---[Note]---\nThe feature References and Incognito mode is available only for the first tab that is created after the Home Page and can be operated from only that tab. Closing the tab will result in error")



        self.show_area.setObjectName("show_area")

        self.History_btn = QtWidgets.QPushButton(self)
        self.History_btn.setGeometry(QtCore.QRect(0, 70, 241, 61))
        font = QtGui.QFont()
        font.setFamily("Cambria")
        font.setPointSize(14)
        self.History_btn.setFont(font)
        self.History_btn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.History_btn.setObjectName("History_btn")


        self.label = QtWidgets.QLabel(self)
        self.label.setGeometry(QtCore.QRect(16, 262, 211, 201))
        font = QtGui.QFont()
        font.setFamily("Bahnschrift Light")
        font.setPointSize(11)
        self.label.setFont(font)
        self.label.setAlignment(QtCore.Qt.AlignCenter)
        self.label.setObjectName("label")

        self.retranslateUi()
        QtCore.QMetaObject.connectSlotsByName(self)

        self.show()


    def retranslateUi(self):
        _translate = QtCore.QCoreApplication.translate
        self.setWindowTitle(_translate("Dialog", "Option"))
        self.Incog_btn.setText(_translate("Dialog", "Turn On Incognito"))
        self.Bookmarks_btn.setText(_translate("Dialog", "Bookmarks"))

        self.History_btn.setText(_translate("Dialog", "History"))
        self.label.setText(_translate("Dialog", "Nox Version 1.0.0\nDate: June 2020\nCreator: Geetanjali Aich"))







Nox.py
import os
import sys
import datetime
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import *
from PyQt5 import QtCore, Qt
from PyQt5.QtCore import (Qt, pyqtSignal,QUrl, QSize)
from PyQt5.QtWebEngineWidgets import *
from option_window_class2 import *

#first installation of PyQt5, QWebEngine, datetime, os, sys is required

#This class is used for creating the tab of the browser

class TabPage(QMainWindow):

    def __init__(self,urlview,t_no):
        super().__init__()
        # First, initialize the tab page with the url it is going to open and the tab number.
        # and set the incognito and reference mode flags down
        self.tab_no=t_no

        self.incog_flag = 0
        self.ref_flag = 0


        # The Tab Page contains three section :
        # 1. The Toolbar section
        # 2. The webview section called the pageview
        # 3. The status bar

        self.url = urlview

        # Setting up the webview section

        self.pagewindow = QWebEngineView()

        self.pagewindow.setUrl(QUrl(urlview))


        # This signal is generated when the url has changed in the current tab page object
        self.pagewindow.urlChanged.connect(self.update_urlbar)
        # This signal is generated when the loading the current url has finished
        self.pagewindow.loadFinished.connect(self.update_title)


        self.setCentralWidget(self.pagewindow)

        # setting up the status bar

        self.status = QStatusBar()
        self.setStatusBar(self.status)

        # setting up the toolbar

        self.navtb = QToolBar("Navigation")
        self.navtb.setIconSize(QSize(32, 32))
        self.addToolBar(self.navtb)

        self.back_btn = QAction(QIcon(os.path.join('images', 'images/back.png')), "Back", self)
        self.back_btn.setStatusTip("Back to previous page")
        self.back_btn.triggered.connect(self.pagewindow.back)
        self.back_btn.setIcon(QtGui.QIcon('images/back.png'))
        self.navtb.addAction(self.back_btn)

        self.next_btn = QAction(QIcon(os.path.join('images', 'images/forward.png')), "Forward", self)
        self.next_btn.setStatusTip("Forward to next page")
        self.next_btn.triggered.connect(self.pagewindow.forward)
        self.next_btn.setIcon(QtGui.QIcon('images/forward.png'))
        self.navtb.addAction(self.next_btn)

        self.reload_btn = QAction(QIcon(os.path.join('images', 'images/refresh.png')), "Reload", self)
        self.reload_btn.setStatusTip("Reload page")
        self.reload_btn.triggered.connect(self.pagewindow.reload)
        self.reload_btn.setIcon(QtGui.QIcon('images/refresh.png'))
        self.navtb.addAction(self.reload_btn)

        self.home_btn = QAction(QIcon(os.path.join('images', 'images/home.png')), "Home", self)
        self.home_btn.setStatusTip("Go home")
        self.home_btn.triggered.connect(self.navigate_home)
        self.home_btn.setIcon(QtGui.QIcon('images/google.png'))
        self.navtb.addAction(self.home_btn)

        self.navtb.addSeparator()

        self.placeholderspace = QLabel()
        self.navtb.addWidget(self.placeholderspace)

        self.urlbar = QLineEdit()
        self.urlbar.returnPressed.connect(self.navigate_to_url)
        self.navtb.addWidget(self.urlbar)


        self.placeholderspace2 = QLabel()
        self.navtb.addWidget(self.placeholderspace2)


        self.navtb.addSeparator()

        self.book_btn = QAction(QIcon(os.path.join('images', 'images/stop.png')), "Bookmark This Page", self)
        self.book_btn.setStatusTip("Bookmark current page")
        self.book_btn.triggered.connect(self.bookmarkSaver)
        self.book_btn.setIcon(QtGui.QIcon('images/bookmark.png'))
        self.navtb.addAction(self.book_btn)

        self.stop_btn = QAction(QIcon(os.path.join('images', 'images/stop.png')), "Stop", self)
        self.stop_btn.setStatusTip("Stop loading current page")
        self.stop_btn.triggered.connect(self.pagewindow.stop)
        self.stop_btn.setIcon(QtGui.QIcon('images/stop.png'))
        self.navtb.addAction(self.stop_btn)

        self.option_btn = QAction(QIcon(os.path.join('images', 'images/stop.png')), "Options", self)
        self.option_btn.setStatusTip("Options")
        self.option_btn.triggered.connect(self.option)
        self.option_btn.setIcon(QtGui.QIcon('images/option.png'))
        self.navtb.addAction(self.option_btn)

        self.show()

    def navigate_home(self):
        '''
        This function is triggered when the home button is clicked.
        '''
        self.pagewindow.setUrl(QUrl("http://www.google.com"))

    def navigate_to_url(self):
        '''
        Useful for setting the url of the pagewindow.
        '''
        q = QUrl(self.urlbar.text())
        if q.scheme() == "":
            q.setScheme("http")
        self.pagewindow.setUrl(q)

    def update_title(self):
        '''
        Updates the title of the tab
        '''
        self.title = self.pagewindow.page().title()

    def update_urlbar(self, q):
        '''
        This function, updates the urlbar with q and at the same time records
        history if incognito is off or reference mode is on. While updating it
        is necessary to check if the page has been bookmarked. If it has been
        bookmarked, then the bookmark button needs to be turned blue(Which is obtained from the images).
        :param q: URL of the page
        :return:
        '''

        self.urlbar.setText(q.toString())
        self.urlbar.setCursorPosition(0)

        if self.incog_flag == 0:
            self.history_writer = open("History.txt", "a")
            now = datetime.datetime.now()
            n=now.strftime("%Y-%m-%d %H:%M:%S")
            self.historyElement="  [ "+n+" ]  "+q.toString()+"\n\n" # Finds the history
            self.history_writer.write(self.historyElement)
            self.history_writer.close()
        if self.ref_flag == 1:
            self.historyElement = q.toString() + "\n\n"
            self.ref_file.write(self.historyElement)

        bm=open("Bookmarks.txt", "r")
        b=bm.readlines()
        l=q.toString()+'\n'
        if l in b:
            self.book_btn.setIcon(QtGui.QIcon('images/bookmark_done.png'))

        else:
            self.book_btn.setIcon(QtGui.QIcon('images/bookmark.png'))
        bm.close()

    def option(self):
        '''
        This function is triggered when the options button is clicked.
        It opens up a options_window_class2 object for the current page.
        :return:
        '''
        self.dlg=Ui_Dialog(self.tab_no)

        #This logic is used for toggling of the incognito and reference button

        if self.ref_flag == 0:
            self.dlg.Ref_btn.setText("Turn ON Reference")
        else:
            self.dlg.Ref_btn.setText("Turn OFF Reference")

        if self.incog_flag == 0:
            self.dlg.Incog_btn.setText("Turn ON Incognito")
        else:
            self.dlg.Incog_btn.setText("Turn OFF Incognito")

        # If any of the buttons are clicked, then appropriate constant will be passed to the
        # WriteOnShowArea function. Accordingly, the necessary text files will be displayed.

        self.dlg.Incog_btn.clicked.connect(lambda: self.WriteOnShowArea("i"))
        self.dlg.Bookmarks_btn.clicked.connect(lambda: self.WriteOnShowArea("b"))
        self.dlg.Ref_btn.clicked.connect(lambda: self.WriteOnShowArea("r"))
        self.dlg.History_btn.clicked.connect(lambda: self.WriteOnShowArea("h"))

        self.dlg.exec_()

    def WriteOnShowArea(self, option):
        '''
        This function will show necessary text files onto the show area.
        :param option: may be h,b,i or r as per the button which is clicked
        :return:
        '''
        self.dlg.show_area.clear()
        if option=="h":
            t = open("History.txt", "r")
            self.dlg.show_area.setText(t.read())
            t.close()
        elif option=="b":
            t = open("Bookmarks.txt", "r")
            self.dlg.show_area.setText(t.read())
            t.close()
        elif option=="i":
            if self.incog_flag == 0:
                self.dlg.show_area.setText('Incognito is now ON\nSearch History will not be stored')
                self.incog_flag = 1
                self.dlg.Incog_btn.setText("Turn OFF Incognito")

            else:
                self.dlg.show_area.setText('Incognito is now OFF')
                self.incog_flag = 0
                self.dlg.Incog_btn.setText("Turn ON Incognito")

        elif option=="r":


            if self.ref_flag == 0:
                self.dlg.show_area.setText('Reference has been turned on')
                self.ref_file=open("References.txt","w")
                self.ref_file.write('\n--------------------------------\n')
                self.dlg.Ref_btn.setText("Turn OFF Reference")
                self.ref_flag = 1
            elif self.ref_flag == 1:
                self.ref_file.close()
                self.ref_file = open("References.txt","r")
                self.dlg.show_area.setText('References:\n'+self.ref_file.read())
                self.ref_file.close()
                os.remove("References.txt")
                self.dlg.Ref_btn.setText("Turn ON Reference")
                self.ref_flag=0

    def bookmarkSaver(self):
        '''
        This function is useful for writing on the bookmark.txt file
        :return:
        '''

        link = self.urlbar.text() + "\n"
        self.book = open("Bookmarks.txt", "r")
        links=self.book.readlines()
        #print(links)
        self.book.close()

        findflag=0
        for l in links:
            if l == link:
                findflag=1
                break

        if findflag == 0:
            self.book_write = open("Bookmarks.txt", "a")
            self.book_write.write(link)
            self.book_write.close()

        self.book_btn.setIcon(QtGui.QIcon('images/bookmark_done.png'))

#This HomeTabWindow object is created when the application is executed

class HomeTabWindow(QWidget):

    def __init__(self):
        super().__init__()

        # This is used for taking care of the tabs
        self.tabcount = 0
        self.tablist = list()

        # Setting up the user interface of the Home Tab
        self.UI = QTabWidget()

        # Setting the size of window
        self.UI.setTabsClosable(True)
        self.UI.setMovable(True)
        self.UI.setTabPosition(QTabWidget.North)
        self.UI.setWindowIcon(QtGui.QIcon('images/nox.png'))
        self.UI.setMinimumSize(1100,900)
        self.UI.setWindowTitle("Nox")

        # Setting Layout of the home page
        self.setUILayout()
        self.UI.tabCloseRequested.connect(self.tabdestructor)

    def setUILayout(self):
        '''
        Used for setting up the entire hometab user interface
        :return:
        '''

        buttonlayout = QGridLayout()
        buttonlayout.setColumnMinimumWidth(0, 100)
        buttonlayout.setColumnMinimumWidth(1, 100)
        buttonlayout.setColumnMinimumWidth(2, 100)

        buttonlayout.setRowMinimumHeight(0, 100)
        buttonlayout.setRowMinimumHeight(1, 100)
        buttonlayout.setRowMinimumHeight(2, 100)

        self.UI.Google = QPushButton(self)
        self.UI.Google.setMaximumSize(200, 200)
        self.UI.Google.setMinimumSize(200, 200)
        self.UI.Google.setObjectName("Google")
        self.UI.Google.setIcon(QtGui.QIcon('images/google.png'))
        self.UI.Google.setIconSize(QtCore.QSize(200, 200))
        self.UI.Google.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.UI.Google.clicked.connect(lambda: self.createTab("https://google.com/"))
        buttonlayout.addWidget(self.UI.Google, 0, 0)

        self.UI.LinkedIn = QPushButton(self)
        self.UI.LinkedIn.setMaximumSize(200, 200)
        self.UI.LinkedIn.setMinimumSize(200, 200)
        self.UI.LinkedIn.setObjectName("LinkedIn")
        self.UI.LinkedIn.setIcon(QtGui.QIcon('images/linkedin.png'))
        self.UI.LinkedIn.setIconSize(QtCore.QSize(200, 200))
        self.UI.LinkedIn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.UI.LinkedIn.clicked.connect(lambda: self.createTab("https://linkedin.com/"))
        buttonlayout.addWidget(self.UI.LinkedIn, 0, 1)

        self.UI.ResearchGate = QPushButton(self)
        self.UI.ResearchGate.setMaximumSize(200, 200)
        self.UI.ResearchGate.setMinimumSize(200, 200)
        self.UI.ResearchGate.setObjectName("ResearchGate")
        self.UI.ResearchGate.setIcon(QtGui.QIcon('images/researchgate.png'))
        self.UI.ResearchGate.setIconSize(QtCore.QSize(200, 200))
        self.UI.ResearchGate.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.UI.ResearchGate.clicked.connect(lambda: self.createTab("https://researchgate.net/"))
        buttonlayout.addWidget(self.UI.ResearchGate, 0, 2)

        self.UI.Youtube = QPushButton(self)
        self.UI.Youtube.setMaximumSize(200, 200)
        self.UI.Youtube.setMinimumSize(200, 200)
        self.UI.Youtube.setObjectName("Youtube")
        self.UI.Youtube.setIcon(QtGui.QIcon('images/youtube.png'))
        self.UI.Youtube.setIconSize(QtCore.QSize(200, 200))
        self.UI.Youtube.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.UI.Youtube.clicked.connect(lambda: self.createTab("https://youtube.com/"))
        buttonlayout.addWidget(self.UI.Youtube, 1, 0)

        self.UI.IEEEXplore = QPushButton(self)
        self.UI.IEEEXplore.setMaximumSize(200, 200)
        self.UI.IEEEXplore.setMinimumSize(200, 200)
        self.UI.IEEEXplore.setObjectName("IEEEXplore")
        self.UI.IEEEXplore.setIcon(QtGui.QIcon('images/ieeexplore.png'))
        self.UI.IEEEXplore.setIconSize(QtCore.QSize(200, 200))
        self.UI.IEEEXplore.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.UI.IEEEXplore.clicked.connect(lambda: self.createTab("https://ieeexplore.ieee.org/"))
        buttonlayout.addWidget(self.UI.IEEEXplore, 1, 1)

        self.UI.Search = QPushButton(self)
        self.UI.Search.setMaximumSize(200, 200)
        self.UI.Search.setMinimumSize(200, 200)
        self.UI.Search.setObjectName("Search")
        self.UI.Search.setIcon(QtGui.QIcon('images/search.png'))
        self.UI.Search.setIconSize(QtCore.QSize(200, 200))
        self.UI.Search.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        buttonlayout.addWidget(self.UI.Search, 1, 2)
        self.UI.Search.clicked.connect(lambda: self.createTab())

        buttonplaceholder = QWidget()
        buttonplaceholder.setLayout(buttonlayout)

        vbox = QVBoxLayout()
        vbox.setAlignment(Qt.AlignCenter)

        self.UI.label = QLabel(self)
        font = QtGui.QFont()
        font.setFamily("Book Antiqua")
        font.setPointSize(45)
        font.setBold(True)
        font.setWeight(75)
        self.UI.label.setFont(font)
        self.UI.label.setCursor(QtGui.QCursor(QtCore.Qt.ArrowCursor))
        self.UI.label.setAlignment(QtCore.Qt.AlignHCenter)
        self.UI.label.setText("Getting Started with\nNox\n")

        placeholdertop = QLabel(self)
        placeholdertop.setText("\n\n")

        vbox.addWidget(placeholdertop)
        vbox.addWidget(self.UI.label)
        vbox.addWidget(buttonplaceholder)

        Home = QWidget()
        Home.setLayout(vbox)

        self.UI.addTab(Home, "Home")

        self.UI.show()

    def createTab(self,starttaburl="http://google.com"):
        '''
        This function created a tab page by passing on the url of the tab to be
        created and the count of the tab.i.e, which tab it is.
        :param starttaburl:
        :return:
        '''

        self.tabcount = self.tabcount + 1

        self.UI.tab = TabPage(starttaburl, len(self.tablist)+1)
        self.tablist.insert(self.tabcount-1, self.UI.tab)
        self.UI.addTab(self.tablist[self.tabcount-1],"Loading...")
        self.UI.setCurrentIndex(self.tabcount)
        self.tablist[self.tabcount-1].pagewindow.loadFinished.connect(self.printer)
        self.UI.setTabText(self.tabcount,"Loading...")
        # If a download request is encountered while tab creation, the dialog box is opened.
        self.UI.tab.pagewindow.page().profile().downloadRequested.connect(self.on_downloadRequested)

    @QtCore.pyqtSlot("QWebEngineDownloadItem*")
    def on_downloadRequested(self, download):

        old_path = download.url().path()  # download.path()
        suffix = QtCore.QFileInfo(old_path).suffix()
        path, _ = QtWidgets.QFileDialog.getSaveFileName(
            self, "Save File", old_path, "*." + suffix
        )
        if path:
            download.setPath(path)
            download.accept()
            download.finished.connect(self.foo)

    def foo(self):
        '''
        This function prints finish on the console when download is completed
        :return:
        '''
        print("finished")

    def tabdestructor(self, index):
        '''
        This method is useful for deleting a tab. Since, one needs to unsubscribe while closing
        a youtube tab, a logic has been added to handle youtube pages. If it is a
        youtube page, then first set the url of the page to google (this will stop the sound)
        and then close it. If the tab is the home tab, then the application is closed.
        :param index:
        :return:
        '''

        if index != 0:
            self.UI.setCurrentIndex(index)

            self.tabcount =self.tabcount-1

            if "youtube" in self.tablist[index-1].url:
                self.tablist[index-1].pagewindow.setUrl(QUrl("https://google.com"))

            self.tablist.pop(index-1)
            self.UI.removeTab(index)
            self.UI.setCurrentIndex(index-1)

        else:
            sys.exit()

    def printer(self):
        '''
        Used to set the tab title once data has been retrieved
        :return:
        '''
        if self.tabcount != 0:
            self.UI.setTabText(self.tabcount,self.UI.tab.title)





if __name__ == "__main__":
    app = QApplication(sys.argv)
    oMainwindow = HomeTabWindow()

    sys.exit(app.exec_())





